artiq/dashboard/moninj.py

```diff
@@ -179,14 +181,81 @@ def sort_key(self):
         raise NotImplementedError
 
 
+class UrukulModel:
+    def __init__(self, cpld, dds_type, ref_clk, pll, clk_div=0):
+        self.cpld = cpld
+        self.cur_frequency_low = 0
+        self.cur_frequency_high = 0
+        self.cur_reg = 0
+        self.dds_type = dds_type
+        self.cur_amp = 0
+
+        if dds_type == "AD9910":
+            max_freq = 1 << 32
+            clk_mult = [4, 1, 2, 4]
+        else:  # AD9912
+            max_freq = 1 << 48
+            clk_mult = [1, 1, 2, 4]
+        sysclk = ref_clk / clk_mult[clk_div] * pll
+        self.ftw_per_hz = 1 / sysclk * max_freq
+
+    def get_frequency(self):
+        return self.cur_frequency_low + self.cur_frequency_high
+
+    def monitor_update(self, probe, value):
+        probe_type = probe // 4
+        if probe_type == 0:  # probes 0-3: register addresses
+            self._update_reg(value)
+        elif probe_type == 1:  # probes 4-7: data_high (for 64 bit transfer)
+            self._update_data_high(value)
+        elif probe_type == 2:  # probes 8-11: data_low (for 64 bit) or just data (32 bit)
+            self._update_data_low(value)
+
+    def _update_reg(self, reg):
+        if self.dds_type == "AD9910":
+            self.cur_reg = (reg >> 24) & 0xff
+        else: # AD9912
+            self.cur_reg = ((reg >> 16) & ~(3 << 13)) & 0xffff
+
+    def _update_data_low(self, value):
+        if self.dds_type == "AD9910":
+            if (AD9910_REG_PROFILE0() <= self.cur_reg <= AD9910_REG_PROFILE7() or
+                    self.cur_reg == AD9910_REG_FTW()):
+                self.cur_frequency_low = self._ftw_to_freq(value)
```

Comment:
Those non-atomic updates aren't nice.
If atomic is difficult then ignore the lower FTW bits on AD9912 (with corresponding loss of resolution) - it doesn't sound worse than those non-atomic updates, and it is simpler.

Reply:
AD9912 values without the low part don't look great:

![Screenshot from 2022-06-22 12-31-22](https://user-images.githubusercontent.com/1278433/174943940-3ec2526c-e7a8-42f1-9784-3901f6c0b196.png)

Atomic updates kind of don't exactly make sense either - sometimes only the high or only the low part would be updated (e.g. from 10.075 to 10.070MHz - only low part will get updated; surely there's a combination in which only the high part updated as well). So, if low part or high part comes by itself - these are both valid options. If there was a way for moninj to indicate end of probe updates for given channel, then that could be used as an indicator for the update - but adding one would be too much for this tiny issue.

Reply:
> AD9912 values without the low part don't look great:

That's surprising - if you ignore the lower 16 FTW bits (or round the result instead of just truncating), you should still get 32 bits of precision and it should be as precise as the AD9910.
Are you sure you are truncating and interpreting the truncated FTW correctly?

Reply:
FYI, the AD9914 driver has a mode with 63-bit FTW, of which the lower 31 bits are ignored by moninj, and it does not have issues like that.

Reply:
> Atomic updates kind of don't exactly make sense either - sometimes only the high or only the low part would be updated (e.g. from 10.075 to 10.070MHz - only low part will get updated; surely there's a combination in which only the high part updated as well)

Transmitting only the changed bytes is a small and unimportant optimization. But if you don't do atomic updates, and the CPU looks at the probes at the wrong time while the FTW bits are completely changed by the RTIO/SPI system, then the frequency display would glitch (as only half the FTW took the new value) and show a completely incorrect value until the next moninj scan.

Truncating the FTW like the AD9914 does (when operated in extended resolution mode) is an easy way out of this problem.